Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,89 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a hook that allows managed code
+to run before the main application's entry point.
+
+## Motivation
+
+This would allow hosting providers to put configuration and policy in
+a single location, including settings that potentially influence load
+behavior of the main entry point such as the AssemblyLoadContext
+behavior. This could be used to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook needs to be separate from the
+entry point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+Environment variables or `<appname>.runtimeconfig.json` can be used to
+specify a managed assembly and type that contains an `Initialize`
+method.
+
+```
+DOTNET_MANAGED_HOST_ASSEMBLY=/path/to/ManagedHost.dll
+DOTNET_MANAGED_HOST_TYPE=ManagedHostNamespace.ManagedHostType
+```
+
+```
+{
+    "runtimeOptions": {
+        "managedHostAssembly": "/path/to/ManagedHost.dll",
+        "managedHostType": "ManagedHostNamespace.ManagedHostType"
+    }
+}
+```
+
+The environment variables, if set, take precedence over the
+`<appname>.runtimeconfig.json` settings. These settings result in
+`ManagedHostType.Initialize()` being called in hostpolicy, before the
+main assembly is loaded.
+
+Specifically, hostpolicy starts up coreclr and sets up a new AppDomain
+with `ManagedHost.dll` on the TPA list. If it read these settings from
+environment variables, it clears them so that child processes do not
+inherit these settings. It then invokes
+`ManagedHost.Initialize()`. This gives `ManagedHost` a chance to set
+up new AssemblyLoadContexts, or register other callbacks. After
+`Initialize()` returns, hostpolicy starts up the main entry point of
+the app like usual.
+
+Rather than forcing all configuration to be done through a single
+predefined API, this creates a place where such configuration could be
+centralized, while still allowing user code to do its own thing if it
+so desires.
+
+The producer of `ManagedHost.dll` needs to ensure that
```

Comment:
ManagedHost.dll has to be built against same or lower version of .NET Core.

Reply:
Pointed this out in the doc more explicitly.

Reply:
Sounds good. What about dependencies to CoreFX? If ManagedHost.dll is built with lower version of CoreFX assemblies and that got unified to the application packaged higher version at runtime, it should be fine.

Reply:
Yup, no unification will happen. We're making no attempt to process a deps file for ManagedHost.dll - it's just an assembly. It needs to be compatible with the dependencies in the *app's* deps.json.

Reply:
Imagine the managedhost.dll in the platform is built with .NET Core 3.0 + CoreFX 3.0, and user want to run .NET Core 4.0 apps, then the dependencies in managedhost.dll with CoreFX 3.0 won't resolve correctly to CoreFX 4.0. So this effectively forces the path to *build*  managedhost.dll as part of the app, or inject it with the .deps to unify with CoreFX 4.0?

Reply:
Refs roll forward, so the .NET Core 3.0 built managedhost.dll will load fine within a .NET Core 4.0 process.  The inverse is not true.

Reply:
If your concern is about breaking changes between major versions, you'll run into the same issue for any .NET Core 4.0 app that tries to run against .NET Core 3.0. I'm definitely not suggesting building it with the app - this will likely cause contention issues where each app and framework wants control over the first managed code to run (see the discussion below). My understanding is that the host provider would have a shared runtime install that supports user apps targeting 2.x or 3.x, but not both if there are breaking changes. To support all apps targeting that major version, a managed host dll should then be compiled against the lowest minor version of that major version (or at least not reference any surface area that doesn't exist in lower minor versions).